contracts/utils/test/log_decoding.ts

```diff
@@ -0,0 +1,76 @@
+import { chaiSetup, provider, txDefaults, web3Wrapper } from '@0x/contracts-test-utils';
+import { BlockchainLifecycle } from '@0x/dev-utils';
+import { BigNumber } from '@0x/utils';
+import * as chai from 'chai';
+import { DecodedLogArgs, LogWithDecodedArgs } from 'ethereum-types';
+
+import { artifacts, TestLogDecodingContract } from '../src';
+
+chaiSetup.configure();
+const expect = chai.expect;
+
+const blockchainLifecycle = new BlockchainLifecycle(web3Wrapper);
+
+describe('TestLogDecoding', () => {
+    let testLogDecodingWithDependencies: TestLogDecodingContract;
+    let testLogDecodingDeployedWithoutDependencies: TestLogDecodingContract;
+    const expectedEvent = {
+        foo: new BigNumber(256),
+        bar: '0x1234',
+        car: '4321',
+    };
+    const expectedDownstreamEvent = {
+        lorem: new BigNumber(256),
+        ipsum: '4321',
+    };
+    const emptyDependencyList = {};
+
+    before(async () => {
+        testLogDecodingDeployedWithoutDependencies = await TestLogDecodingContract.deployFrom0xArtifactAsync(
+            artifacts.TestLogDecoding,
+            provider,
+            txDefaults,
+            emptyDependencyList,
+        );
+        testLogDecodingWithDependencies = await TestLogDecodingContract.deployFrom0xArtifactAsync(
+            artifacts.TestLogDecoding,
+            provider,
+            txDefaults,
+            artifacts,
+        );
+    });
+    beforeEach(async () => {
+        await blockchainLifecycle.startAsync();
+    });
+    afterEach(async () => {
+        await blockchainLifecycle.revertAsync();
+    });
+
+    describe('Decoding Log Arguments', () => {
```

Comment:
It isn't immediately clear from the tests what the behavior is when we do an `emitEvent()` and `emitEventDownstream()` at once, without downstream artifacts. I assume `TestEvent` will turn into a `DecodedLogEntry` and `TestEvent2` will remain a `LogEntry`, but it might be useful to demonstrate it.

Reply:
Heh I actually had a test for that and removed it. I'll put it back in. 🙏